File: src/Lab4/GUI.java
package Lab4;

import Graph.MatrixGraph;

import javax.imageio.ImageIO;
import javax.swing.*;
import javax.swing.filechooser.FileNameExtensionFilter;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.image.BufferedImage;
import java.io.*;
import java.util.Scanner;

public class GUI extends JFrame implements ActionListener {

    private JPanel buttonPanel, argsPanel, resultPanel;
    private JButton open, reset, dijkstra;
    private JTextField from, to;
    private JLabel result;

    private int[] path;

    public GUI() {
        setTitle("Dijkstra");
        setDefaultCloseOperation(EXIT_ON_CLOSE);
        setSize(350, 450);
        setResizable(false);
        setLayout(new GridLayout(3, 1));
        buttonPanel = new JPanel();
        argsPanel = new JPanel();
        resultPanel = new JPanel();
        add(buttonPanel);
        add(argsPanel);
        add(resultPanel);
        initButtonPanel();
        initArgsPanel();
        initResultPanel();
    }

    private void initButtonPanel() {
        buttonPanel.setLayout(null);
        open = new JButton("OPEN");
        reset = new JButton("RESET");
        reset.setEnabled(false);
        open.addActionListener(this);
        reset.addActionListener(this);
        open.setBounds(50, 50, 100, 50);
        reset.setBounds(200, 50, 100, 50);
        buttonPanel.setSize(350, 150);
        buttonPanel.add(open);
        buttonPanel.add(reset);
    }

    private void initArgsPanel() {
        argsPanel.setLayout(null);
        JLabel l1 = new JLabel("From:");
        JLabel l2 = new JLabel("To:");
        l1.setBounds(30, 30, 70, 30);
        l2.setBounds(30, 90, 370, 30);
        argsPanel.add(l1);
        argsPanel.add(l2);
        from = new JTextField();
        to = new JTextField();
        from.setBounds(100, 30, 70, 30);
        to.setBounds(100, 90, 70, 30);
        argsPanel.add(from);
        argsPanel.add(to);
        dijkstra = new JButton("DIJKSTRA");
        dijkstra.setEnabled(false);
        dijkstra.addActionListener(this);
        dijkstra.setBounds(220, 30, 90, 90);
        argsPanel.add(dijkstra);
    }

    private void initResultPanel() {
        resultPanel.setLayout(null);
        JLabel msg = new JLabel("Shortest Path Length:");
        msg.setBounds(30, 50, 150, 30);
        result = new JLabel();
        result.setBounds(180, 50, 100, 30);
        resultPanel.add(msg);
        resultPanel.add(result);
    }


    @Override
    public void actionPerformed(ActionEvent e) {
        try {
            if (e.getSource() == open) {
                open();
            } else if (e.getSource() == reset) {
                reset();
            } else if (e.getSource() == dijkstra) {
                dijkstra();
            }
        } catch (IOException ex) {
            ex.printStackTrace();
        }

    }

    private MatrixGraph graph;
    private JFileChooser fileChooser;
    private ShowFrame showFrame;

    private void open() throws IOException {
        if (fileChooser == null) {
            fileChooser = new JFileChooser();
            fileChooser.setDialogTitle("Select The File to Build Graph");
            fileChooser.setFileFilter(new FileNameExtensionFilter("TEXT File", "txt"));
            fileChooser.setAcceptAllFileFilterUsed(false);
            fileChooser.setMultiSelectionEnabled(false);
        }
        fileChooser.showOpenDialog(GUI.this);
        var fin = fileChooser.getSelectedFile();
        if (fin != null) {
            Scanner sc = new Scanner(new BufferedReader(new FileReader(fin)));
            int n = sc.nextInt();
            var vexs = new String[n];
            for (int i = 0; i < n; i++) {
                vexs[i] = "V" + i;
            }
            var arcs = new int[n][n];
            for (int i = 0; i < n; i++) {
                arcs[i][i] = 0;
                for (int j = 0; j < n; j++) {
                    arcs[i][j] = sc.nextInt();
                }
                for (int j = 0; j < i; j++) {
                    arcs[i][j] = arcs[j][i];
                }
            }
            graph = new MatrixGraph(vexs, arcs);
            graph.resetLastPath();
            reset.setEnabled(true);
            dijkstra.setEnabled(true);
            showGraph();
        }
    }

    private void reset() throws IOException {
        graph.resetLastPath();
        showGraph();
    }

    private void dijkstra() throws IOException {
        int a = Integer.valueOf(from.getText());
        int b = Integer.valueOf(to.getText());
        if (a < 0 || a >= graph.getVexNumber() || b < 0 || b >= graph.getVexNumber()) {
            //非法输入
            JOptionPane.showConfirmDialog(GUI.this,
                    "Please input right start anf end vex.",
                    "Illegal Input!", JOptionPane.OK_OPTION);
            return;
        }
        path = graph.Dijkstra(a, b);
        if (path == null) {
            graph.resetLastPath();
            result.setText("Not Connected");
        } else {
            int len = 0;
            var arcs = graph.getArcs();
            for (int i = 0; i < path.length - 1; i++) {
                len += arcs[path[i]][path[i + 1]];
            }
            result.setText(String.valueOf(len));
        }
        showGraph();
    }

    private void showGraph() throws IOException {
        var str = graph.toDotLanguage();
        var temp_dir = System.getProperty("java.io.tmpdir");
        File dot = new File(temp_dir + "\\graphviz.dot");
        if (dot.exists()) {
            dot.delete();
        }
        var out = new FileWriter(dot);
        out.write(str);
        out.close();

        String cmd[] = {"dot", "-T", "png", "-Gdpi=100", "-o", temp_dir + "\\graphviz.png", dot.getAbsolutePath()};
        try {
            Runtime.getRuntime().exec(cmd).waitFor();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        //显示图片
        if (showFrame == null) {
            showFrame = new ShowFrame("Graph");
        }
        showFrame.setImg(temp_dir + "\\graphviz.png");
        showFrame.setVisible(true);
    }

}

class ShowFrame extends JFrame {
    public ShowFrame(String title) {
        super(title);
        setDefaultCloseOperation(HIDE_ON_CLOSE);
    }

    private BufferedImage img;

    public void setImg(String fileName) throws IOException {
        img = ImageIO.read(new File(fileName));
        setSize(img.getWidth() + 14, img.getHeight() + 39);
        repaint();
    }

    @Override
    public void paint(Graphics g) {
        super.paint(g);
        g.drawImage(img, 7, 30, null);
    }
}

File: src/Lab4/GUI.kt
package Lab4

import Graph.MatrixGraph
import java.awt.Graphics
import java.awt.GridLayout
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.image.BufferedImage
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.util.*
import javax.imageio.ImageIO
import javax.swing.*
import javax.swing.filechooser.FileNameExtensionFilter

fun main(args: Array<String>) {
    GUI().isVisible = true
}

class GUI : JFrame("dijkstra"), ActionListener {

    private val open = JButton("Open")
    private val reset = JButton("Reset")
    private val dijkstra = JButton("dijkstra")

    private val fileChooser = JFileChooser()
    private val showFrame = ShowFrame()

    private val from = JTextField()
    private val to = JTextField()
    private val result = JLabel()

    private lateinit var graph: MatrixGraph

    private lateinit var path: IntArray

    init {
        defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
        setSize(350, 450)
        isResizable = false
        layout = GridLayout(3, 1)

        //初始化按钮面板
        add(JPanel(null).apply {
            add(open.apply {
                addActionListener(this@GUI)
                setBounds(50, 50, 100, 50)
            })
            add(reset.apply {
                isEnabled = false
                addActionListener(this@GUI)
                setBounds(200, 50, 100, 50)
            })
        })

        //参数输入面板
        add(JPanel(null).apply {
            add(JLabel("From:").apply {
                setBounds(30, 30, 70, 30)
            })
            add(JLabel("To:").apply {
                setBounds(30, 90, 370, 30)
            })
            add(from.apply {
                setBounds(100, 30, 70, 30)
            })
            add(to.apply {
                setBounds(100, 90, 70, 30)
            })
            add(dijkstra.apply {
                isEnabled = false
                addActionListener(this@GUI)
                setBounds(220, 30, 90, 90)
            })
        })

        //结果显示面板
        add(JPanel(null).apply {
            add(JLabel("Shortest Path Length:").apply {
                setBounds(30, 50, 150, 30)
            })
            add(result.apply {
                setBounds(180, 50, 100, 30)
            })
        })

        fileChooser.apply {
            dialogTitle = "Select The File to Build Graph"
            fileFilter = FileNameExtensionFilter("Text File", "txt")
            isAcceptAllFileFilterUsed = false
            isMultiSelectionEnabled = false
        }
    }

    override fun actionPerformed(e: ActionEvent) {
        when (e.source) {
            open -> open()
            reset -> reset()
            dijkstra -> dijkstra()
        }
    }

    private fun open() {

        val fin = fileChooser.apply {
            showOpenDialog(this@GUI)
        }.selectedFile ?: return

        //fin是储存信息的文件
        val sc = Scanner(BufferedReader(FileReader(fin)))
        //顶数量
        val n = sc.nextInt()
        //顶信息，使用默认命名
        val vexs = Array(n) { i -> "V$i" }
        //边长度
        val arcs = Array(n) { IntArray(n) }
        for (i in 0 until n) {
            //读取矩阵
            for (j in 0 until n) {
                arcs[i][j] = sc.nextInt()
            }
            //对角化
            for (j in 0 until i) {
                arcs[i][j] = arcs[j][i]
            }
        }
        graph = MatrixGraph(vexs, arcs)

        reset.isEnabled = true
        dijkstra.isEnabled = true
        showGraph()

    }

    private fun reset() {
        graph.resetLastPath()
        showGraph()
    }

    private fun dijkstra() {
        val a = from.text.toInt()
        val b = to.text.toInt()

        if (a in 0 until graph.getVexNumber() && b in 0 until graph.getVexNumber()) {
            path = graph.dijkstra(a, b)
            if (path.isEmpty()) {
                result.text = "Not Connected"
            } else {
                var len = 0
                val arcs = graph.arcs
                for (i in 0 until path.size - 1) {
                    len += arcs[path[i]][path[i + 1]]
                }
                result.text = len.toString()
            }
            showGraph()
        } else {
            //非法输入
            JOptionPane.showConfirmDialog(this@GUI,
                    "Please input right start and end vertexs.",
                    "Illegal Input!", JOptionPane.YES_OPTION)
        }

    }

    private fun showGraph() {
        val tempdir = System.getProperty("java.io.tmpdir")

        FileWriter("$tempdir\\graphviz.dot").apply {
            write(graph.toDotLanguage())
            close()
        }

        //生成png文件，写入到系统临时目录下
        Runtime.getRuntime().exec(
                arrayOf("dot", "-T", "png", "-Gdpi=100",
                        "-o", "$tempdir\\graphviz.png",
                        "$tempdir\\graphviz.dot")
        ).waitFor()

        //显示图片
        showFrame.apply {
            isVisible = true
            setImg("$tempdir\\graphviz.png")
        }
    }

}

internal class ShowFrame : JFrame("Graph") {

    private var img: BufferedImage? = null

    init {
        defaultCloseOperation = WindowConstants.HIDE_ON_CLOSE
    }

    fun setImg(fileName: String) {
        img = ImageIO.read(File(fileName))
        setSize(img!!.width + 14, img!!.height + 39)
        repaint()
    }

    override fun paint(g: Graphics) {
        super.paint(g)
        g.drawImage(img, 7, 30, null)
    }
}
